File: competitive_programming/leetcode/python/212_word_search_ii.py
from typing import List
from typing import Tuple

class Solution:
    def getNeighbors(self, i: int, j: int, valid: List[List[bool]]) -> List[Tuple]:
        neighbors = []
        if i - 1 >= 0 and valid[i - 1][j]:
            neighbors.append((i - 1, j))
        if j - 1 >= 0 and valid[i][j - 1]:
            neighbors.append((i, j - 1))
        if i + 1 < len(valid) and valid[i + 1][j]:
            neighbors.append((i + 1, j))
        if j + 1 < len(valid[0]) and valid[i][j + 1]:
            neighbors.append((i, j + 1))                
        return neighbors
    
    def dfs(self, word: str, row: int, col: int, \
            valid: List[List[bool]], board: List[List[str]]):
        if len(word) == 0:
            return True
        neighbors = self.getNeighbors(row, col, valid)
        for neighbor in neighbors:
            i, j = neighbor
            if board[i][j] == word[0]:
                valid[i][j] = False
                ans = self.dfs(word[1:], i, j, valid, board)
                valid[i][j] = True   
                if ans:
                    return True
        return False

    def findWord(self, valid: List[List[str]], board: List[List[str]], word: str) -> bool:
        m = len(board)
        n = len(board[0])
        for i in range(m):
            for j in range(n):
                if board[i][j] == word[0]:
                    valid[i][j] = False
                    ans = self.dfs(word[1:], i, j, valid, board)
                    valid[i][j] = True
                    if ans:
                        return True
        return False

    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        valid = [[True for _ in board[0]] for _ in board]
        ans = []
        for word in words:
            if self.findWord(valid, board, word):
                ans.append(word)
        return ans

solution = Solution()

board = [ \
    ["o","a","a","n"], \
    ["e","t","a","e"], \
    ["i","h","k","r"], \
    ["i","f","l","v"]  \
]
words = ["oath","pea","eat","rain"]
print(solution.findWords(board, words))

board = [["a","b"],["c","d"]]
words = ["abcb"]
print(solution.findWords(board, words))

File: competitive_programming/leetcode/cpp/212_word_search_ii.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

bool dfs(int row, int col, int wi, vector<vector<char>>& board, int m, int n, string& word) {
    if (wi == word.size()) {
        return true;
    }
    if (row < 0 || row >= m || col < 0 || col >= n || board[row][col] != word[wi]) {
        return false;
    }
    char cur = board[row][col];
    board[row][col] = 0;
    bool ans = dfs(row - 1, col, wi + 1, board, m, n, word) || 
        dfs(row + 1, col, wi + 1, board, m, n, word) ||
        dfs(row, col - 1, wi + 1, board, m, n, word) ||
        dfs(row, col + 1, wi + 1, board, m, n, word);
    board[row][col] = cur;
    if (ans) {
        return true;
    }
    return false;
}

bool findWord(vector<vector<char>>& board, string& word, int m ,int n) {
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            if (dfs(i, j, 0, board, m, n, word)) {
                return true;
            }
        }
    }
    return false;
}

vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
    vector<string> ans;
    int m = board.size();
    int n = board[0].size();
    for (string word: words) {
        if (findWord(board, word, m, n)) {
            ans.push_back(word);
        }
    }
    return ans;
}

void print(vector<string> vec) {
    cout << "--------\n";
    for (string str : vec) {
        cout << str << "\n";
    }
}

int main() {
    vector<string> ans;

    vector<vector<char>> board = {
        {'o','a','a','n'},
        {'e','t','a','e'},
        {'i','h','k','r'},
        {'i','f','l','v'}
    };
    vector<string> words = {"eat", "oath"};
    ans = findWords(board, words);
    print(ans);

    return 0;
}